jupyterhub/templates/proxy/autohttps/_configmap-dynamic.yaml

```diff
@@ -0,0 +1,108 @@
+{{- define "jupyterhub.dynamic.yaml" -}}
+# Content of dynamic.yaml to be merged merged with
+# proxy.traefik.extraDynamicConfig.
+# ----------------------------------------------------------------------------
+http:
+  # Middlewares tweaks requests. We define them here and reference them in
+  # our routers. We use them to redirect http traffic and headers to proxied
+  # web requests.
+  #
+  # ref: https://docs.traefik.io/middlewares/overview/
+  middlewares:
+    hsts:
+      # A middleware to add a HTTP Strict-Transport-Security (HSTS) response
+      # header, they function as a request for browsers to enforce HTTPS on
+      # their end in for a given time into the future, and optionally
+      # subdomains for requests to subdomains as well.
+      #
+      # ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security
+      headers:
+        stsIncludeSubdomains: true
+        stsSeconds: {{ .Values.proxy.traefik.hsts.maxAge | int64 }}
+    # A middleware to redirect to https
+    redirect:
+      redirectScheme:
+        scheme: https
+    # A middleware to add a X-Scheme (X-Forwarded-Proto) header that
+    # JupyterHub's Tornado web-server needs if expecting to serve https
+    # traffic. Without it we would run into issues like:
+    # https://github.com/jupyterhub/jupyterhub/issues/2284
+    scheme:
+      headers:
+        customRequestHeaders:
+          # DISCUSS ME: Can we use the X-Forwarded-Proto header instead? It
+          # seems more recognized. Mozilla calls it the de-facto standard
+          # header for this purpose, and Tornado recognizes both.
+          #
+          # ref: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Proto
+          # ref: https://www.tornadoweb.org/en/stable/httpserver.html#http-server
+          X-Scheme: https
+
+  # Routers routes web requests to a service and optionally tweaks them with
+  # middleware.
+  #
+  # ref: https://docs.traefik.io/routing/routers/
+  routers:
+    # Route secure https traffic to the configurable-http-proxy managed by
+    # JupyterHub.
+    default:
+      entrypoints:
+        - "https"
+      middlewares:
+        - "hsts"
+        - "scheme"
+      rule: PathPrefix(`/`)
+      service: default
+      # Use our predefined TLS options and certificate resolver, enabling
+      # this route to act as a TLS termination proxy with high security
+      # standards.
+      tls:
+        certResolver: default
+        domains:
+          {{- range $host := .Values.proxy.https.hosts }}
+          - main: {{ $host }}
+          {{- end }}
+        options: default
+
+    # Route insecure http traffic to https
+    insecure:
+      entrypoints:
+        - "http"
+      middlewares:
+        - "hsts"
```

Comment:
I had to lookup HSTS to better think of this, and I conclude that HSTS is a header sent back to the browser to indicate it should always use HTTPS and never accept HTTP when connecting with this domain, for a given period of time at least. Such instruction could make sense to pass during any kind of response, even if it is a 302 redirect to https.

Hmmm, but I'm not confident it will be problem free to start demanding that from there already, so perhaps it makes more sense to only have it on HTTPS so that this setup doesn't force HTTPS before the first successful https response.

Hmm.. Remove the hsts middleware on the http->https redirect, which applies the hsts headers?

Reply:
I digged down!

From [the HSTS RFC it seems like](https://tools.ietf.org/html/rfc6797#section-7.2):
1. We "SHOULD" make our redirect from http to https a permanent one.
2. We "MUST NOT"

Nice catch @GeorgianaElena !! :heart: :tada:

Reply:
Thank you for your review @GeorgianaElena! I very often go away happier from reviews by feeling that the code have withstood additional inspect, but the best feeling comes from when it doesn't but it then became better, and then I also often have learned something like now :D